File: python/crusher.py
# Predetermined recycling station. Do not change this file,
# it is outside of your control

from units import CubicMeter
from garbage import Glass, Garbage, GarbageProcessingException


class Crusher:
    def wash(self, volume: CubicMeter):
        print(f"Side effect: washing of {volume} m^3")

    def crush(self, bucket: Garbage):
        if not isinstance(bucket, Glass):
            raise GarbageProcessingException("Attempted to crush non-glass")
        print(f"Side effect: crushing of {bucket} performed")

File: cpp/crusher.cpp
// Predetermined recycling station. Do not change this file,
// it is outside of your control
#include "crusher.h"
#include "garbage.h"

void Crusher::wash(CubicMeter volume) {
    printf("Side effect: washing of %d m^3\n", volume);
}

void Crusher::crush(Garbage *bucket) {
    if (!dynamic_cast<Glass*>(bucket)) {
        throw GarbageProcessingException("Attempted to crush non-glass");
    }
    printf("Side effect: crushing of %p performed\n", bucket);
    }
